kbfsedits/tlf_history.go

```diff
@@ -0,0 +1,354 @@
+			// We have enough edits for this user.
+			return true
+		}
+	case NotificationRename:
+		// If the file already has a final event, move that to the old
+		// filename.  Otherwise, this is the final event.
+		if hasEvent {
+			r.fileEvents[notification.Params.OldFilename] = event
+			delete(r.fileEvents, eventFilename)
+		} else {
+			r.fileEvents[notification.Params.OldFilename] =
+				fileEvent{newName: eventFilename}
+		}
+
+		// The renamed file overwrote any existing file with the new
+		// name.
+		r.fileEvents[eventFilename] = fileEvent{delete: true}
+	case NotificationDelete:
+		r.fileEvents[eventFilename] = fileEvent{delete: true}
+	}
+	return false
+}
+
+func (th *TlfHistory) recomputeLocked() (
+	history writersByRevision, writersWhoNeedMore map[string]bool) {
+	writersWhoNeedMore = make(map[string]bool)
+
+	// Copy the writer notifications into a heap.
+	var writersHeap writersByRevision
+	for _, wn := range th.byWriter {
+		wnCopy := writerNotifications{
+			writerName:    wn.writerName,
+			notifications: make(notificationsByRevision, len(wn.notifications)),
+		}
+		copy(wnCopy.notifications, wn.notifications)
+		writersHeap = append(writersHeap, &wnCopy)
+	}
+	heap.Init(&writersHeap)
+
+	r := newRecomputer()
+
+	// Iterate through the heap.  The writer with the next highest
+	// revision will always be at index 0.  Process that writer's
+	// first notification, then remove it and fix the heap so that the
+	// next highest revision is at index 0.  That way events that
+	// happen more recently (like deletes and renames) can be taken
+	// into account when looking at older events.
+	for writersHeap.Len() > 0 {
+		nextWriter := writersHeap[0].writerName
+		nextNotification := writersHeap[0].notifications[0]
+		doTrim := r.processNotification(nextWriter, nextNotification)
+
+		// Remove that notification, and fix the heap because this
+		// writer has a different newest revision.
+		if doTrim {
+			// Trim all earlier revisions because they won't be needed
+			// for the cached history.
+			numProcessed := r.numProcessed[nextWriter]
+			th.byWriter[nextWriter].notifications =
+				th.byWriter[nextWriter].notifications[numProcessed:]
+		} else {
+			writersHeap[0].notifications = writersHeap[0].notifications[1:]
+		}
+		if len(writersHeap[0].notifications) == 0 || doTrim {
+			heap.Pop(&writersHeap)
+		} else {
+			heap.Fix(&writersHeap, 0)
+		}
+	}
+
+	history = make(writersByRevision, 0, len(r.byWriter))
+	for writerName := range th.byWriter {
+		wn := r.byWriter[writerName]
+		if wn != nil && len(wn.notifications) > 0 {
+			history = append(history, wn)
+		}
+		if wn == nil || len(wn.notifications) < maxEditsPerWriter {
+			writersWhoNeedMore[writerName] = true
+		}
+	}
+	sort.Sort(history)
+	// Garbage-collect any writers that don't appear in the history.
+	for i := maxWritersPerHistory; i < len(history); i++ {
+		delete(th.byWriter, history[i].writerName)
+		delete(writersWhoNeedMore, history[i].writerName)
+	}
```

Comment:
It might be better to move this `for` block into the `if` block below, not for performance reasons but for readability. It took me a while to associate them :)